ops/model.py

```diff
@@ -1326,6 +1327,145 @@ def list_files(self, path: str, *, pattern: str = None,
         """
         return self._pebble.list_files(path, pattern=pattern, itself=itself)
 
+    def push_path(self,
+                  source_path: typing.Union[typing.Union[str,
+                                                         Path],
+                                            typing.List[typing.Union[str,
+                                                                     Path]]],
+                  dest_dir: typing.Union[str,
+                                         Path]):
+        """Push all the things.
+
+        TODO: write this docstring
+        """
+        source_paths = source_path
+        if not isinstance(source_paths, list):
+            source_paths = [source_paths]
+        source_paths = [str(p) for p in source_paths]
+        dest_dir = str(dest_dir)
+
+        def local_list(source_path: str):
+            paths = [source_path]
+            if os.path.isdir(source_path):
+                paths = os.listdir(source_path)
+            files = []
+            for name in paths:
+                fpath = os.path.join(source_path, name)
+                info = os.lstat(fpath)
+                if stat.S_ISDIR(info.st_mode):
+                    ftype = pebble.FileType.DIRECTORY
+                elif stat.S_ISREG(info.st_mode):
+                    ftype = pebble.FileType.FILE
+                elif stat.S_ISLINK(info.st_mode):
+                    ftype = pebble.FileType.SYMLINK
+                else:
+                    continue
+
+                if os.name == 'nt':
+                    files.append(
+                        pebble.FileInfo(
+                            fpath, os.path.basename(fpath), ftype, info.st_size, stat.S_IMODE(
+                                info.st_mode), info.st_mtime))
+                else:
+                    import grp, pwd
+                    files.append(
+                        pebble.FileInfo(
+                            fpath,
+                            os.path.basename(fpath),
+                            ftype,
+                            info.st_size,
+                            stat.S_IMODE(info.st_mode),
+                            info.st_mtime,
+                            info.st_uid,
+                            pwd.getpwuid(info.st_uid).pw_name,
+                            info.st_gid,
+                            grp.getgrgid(info.st_gid).gr_name))
```

Comment:
Edge case, but do we consider the overlay group/pwd from, say, `fat32` filesystems reliable enough to push, or should we drill down to the mount and decide that some subset of filesystems should operate like WIndows files, and not add the uid/gid info?

Reply:
For fat mounts (on linux) these will just give the mount-time-configured user/group for the partition (which defaults to the user making the mount).  This would likely result in us getting 0/'root' for the user sometimes.  If we think this could be a realistic problem in practice, then it wouldn't be too hard to write a little function to check for files being in fat mounts.  I'm just not sure how we envision windows support going forward since there are certainly lots of other things that already don't work quite right or at all on windows.